### PR TITLE
change the way we provide additional params via environment and config file

### LINKED ISSUE #5878: Validator Options parameters not read from `--config-file` when other params are specified via CLI args
Specifying via CLI an option from `ValidatorProposerOptions` (ie `--validators-proposer-config`)  or `ValidatorOptions` (ie `--validators-graffiti`) prevents params from `ValidatorKeysOptions` to be read via yaml config file specified via `--config-file`.

### Example

CLI:
```bash
./teku --config-file ./test.yaml --validators-graffiti test
```
test.yaml
```yaml
validator-keys: ["/etc/teku/keys:/etc/teku/secrets"]
```

`validator-keys` param is ignored and teku will run as it has been never specified.

by removing `--validators-graffiti test` from CLI, `validator-keys` will be read again.

### Initial thought
it seems it's related to how picocli initializes `ArgGroup` groups and how it will apply defaults values.
Since we provide config file functionality as a default value provider, it seems that specifying an argument from a `ArgGroup` causes initialization of all its sub `ArgGroup` which then are considered initialized and no default values are provided.
The same happens if you initialize a sibling group.

A solution could be to flatten all the arguments to `ValidatorOptions` but then we have rethink how to avoid duplication in `VoluntaryExitCommand`.

### PR DESCRIPTION
**Pull request by tbenr**
This PR moves away from relying on picocli default provider and programmatically reconstructs the command line arguments with the additional params from environment and config file.

The basic idea is:
- get all the options available for the current commandLine (or subcommand)
- exclude the options that has been already specified as argument
- search for matching params in environment\configFile
- rebuild the CLI arguments by appending matched params to the original params

fixes #5878

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: teku/src/main/java/tech/pegasys/teku/cli/BeaconNodeCommand.java [hunk 1]
```diff
@@ -216,15 +222,48 @@ public int parse(final String[] args) {
...
+
+    final ParseResult parseResult;
+    try {
+      parseResult = commandLine.parseArgs(args);
+    } catch (ParameterException e) {
+      return handleParseException(e, args);
+    }
+
+    // switch to last subcommand if needed
+    final ParseResult maybeSubCommandParseResult =
```
Comment (ajsutton) [inline teku/src/main/java/tech/pegasys/teku/cli/BeaconNodeCommand.java]: nit: The maybe in the name here keeps making me think it's an `Optional`.  Maybe we should just reuse the `parseResult` field here and reassign it? Or just find a name without the maybe.

File: teku/src/main/java/tech/pegasys/teku/cli/util/EnvironmentVariableParamsProvider.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  @Override
+  public Map<String, String> getAdditionalParams(List<OptionSpec> potentialParams) {
+    return environment.entrySet().stream()
+        .filter(envEntry -> envEntry.getKey().startsWith(ENV_VAR_PREFIX))
+        .map(envEntry -> Map.entry(translateKey(envEntry.getKey()), envEntry.getValue()))
+        .map(yamlEntry -> mapParam(potentialParams, yamlEntry))
+        .filter(Optional::isPresent)
+        .map(Optional::get)
```
Comment (ajsutton) [inline teku/src/main/java/tech/pegasys/teku/cli/util/EnvironmentVariableParamsProvider.java]: Could this just be?
```suggestion
        .flatMap(yamlEntry -> mapParam(potentialParams, yamlEntry).stream())
```

File: teku/src/main/java/tech/pegasys/teku/cli/util/YamlConfigFileParamsProvider.java [hunk 3]
```diff
@@ -26,41 +26,83 @@
...
     }
 
-    // only options can be used in config because a name is needed for the key
-    // so we skip default for positional params
-    return argSpec.isOption() ? getConfigurationValue(((OptionSpec) argSpec)) : null;
+    return result.entrySet().stream()
+        .map(yamlEntry -> Map.entry(translateKey(yamlEntry.getKey()), yamlEntry.getValue()))
+        .map(yamlEntry -> mapParam(potentialParams, yamlEntry))
+        .filter(Optional::isPresent)
+        .map(Optional::get)
```
Comment (ajsutton) [inline teku/src/main/java/tech/pegasys/teku/cli/util/YamlConfigFileParamsProvider.java]:
```suggestion
        .flatMap(yamlEntry -> mapParam(potentialParams, yamlEntry).stream())
```

File: teku/src/main/java/tech/pegasys/teku/cli/util/YamlConfigFileParamsProvider.java [hunk 4]
```diff
@@ -26,41 +26,83 @@
...
+        .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
+  }
+
+  private String translateKey(final String yamlKey) {
+    return "--" + yamlKey;
```
Comment (ajsutton) [inline teku/src/main/java/tech/pegasys/teku/cli/util/YamlConfigFileParamsProvider.java]: I don't think this correctly handles options that use a single `-` which we have a few of, mostly in subcommands.

Probably also applies to the env var handler too though apparently we never handled it properly there...

**Review by ajsutton — COMMENTED** [review]
Generally looks good, I think we just need to check that option names with a single `-` still work in the Yaml provider (and probably fix them in the env provider).

File: teku/src/main/java/tech/pegasys/teku/cli/util/YamlConfigFileParamsProvider.java (hunk 4 above)
Comment (tbenr) [reply teku/src/main/java/tech/pegasys/teku/cli/util/YamlConfigFileParamsProvider.java]: right, we were supporting it in config but not in env.
fixing it.

**Comment by tbenr** [comment]
@ajsutton I managed the situation where multiple config\env referring to the same param are specified.
ie `TEKU_ETH1-ENDPOINT` and `TEKU_ETH1-ENDPOINTS`
Currently I explicitly stop and throw an exception instead of trying to resolve the conflict.
It makes me think that this could be a breaking change because in our previous version we were giving the precedence based on the ordering of `names` in the `@Option` definition.

To print a warning instead and replicate exactly the previous behaviour I should stream starting from Option names... should be doable...

yes I can definitely do the other way around.

**Comment by ajsutton** [comment]
hmm, that is an interesting one.  It's probably not a particularly big deal if we can't support both being set but if we can keep it working that would definitely be better.

**Comment by tbenr** [comment]
done.
Now it should be OK.
The only edge case I can reason about is:
giving an option like this:
```java
  @Option(
      names = {"-n", "--name", "--names"},
      split = ",")
  List<String> getNames();
```
and this env vars
```shell
TEKU_N=a
TEKU_NAME=b
TEKU_NAMES=c
```

previous behaviour: `name` will be `b`
new behaviour: `name` will be `a`

(because now we consider single-dash options too)

Similarly for yaml config file.

**Comment by ajsutton** [comment]
> (because now we consider single-dash options too)

oh darn, I think I've given you bad advice here. I think we deliberately ignored single `-` options in the past because we had problems where teku wouldn't start because someone had an env var set called `TEKU_V` or `TEKU_H` and it triggered the version or help output.  It _could_ potentially have been avoided by only considering options that accept a value (even if it's optional) but I'm not sure which way we went now.

**Comment by ajsutton** [comment]
The extractAPISpec failure is interesting too:
```
Invalid value for option '--logging': 'en_US.UTF-8' is not a valid log level. Supported values are [OFF|FATAL|WARN|INFO|DEBUG|TRACE|ALL]
```

Did we pick up an env var we didn't previously?

**Comment by tbenr** [comment]
```
  @Option(
      names = {"-l", "--logging"}
```
better to filter out again single dashed options...
there must be a `TEKU_L`... let me check... oh no, I know what is happening.

was just removing 5 chars at the begining instead of filtering for `TEKU_`

image

`LC_ALL` was matching `-l`

**Comment by tbenr** [comment]
I reintroduced the `startsWith("--")` requirement on the names

A quick assessment:
image

plus `-h` and `-V`

Now this version should behave exactly as the current one.

**Review by ajsutton — APPROVED** [review]
LGTM. Sorry for the wild goose chase.  This is great work.

**Pull request merged**